opentracing-kotlin/src/main/kotlin/org/zalando/opentracing/kotlin/TracerExtension.kt

```diff
@@ -0,0 +1,94 @@
+package org.zalando.opentracing.kotlin
+
+import io.opentracing.Scope
+import io.opentracing.Span
+import io.opentracing.SpanContext
+import io.opentracing.Tracer
+import io.opentracing.propagation.Format.Builtin.TEXT_MAP
+import io.opentracing.propagation.TextMapAdapter
+import io.opentracing.tag.Tags
+
+inline fun <R> Tracer.trace(span: Span, block: (Span) -> R): R {
+
+    var scope: Scope? = null
+    var closed = false
+
+    try {
+        scope = scopeManager().activate(span)
+
+        return block(span)
+    } catch (e: Exception) {
+        span.record(e)
+        try {
+            scope?.close()
+            closed = true
+        } catch (closeException: Exception) {
+            // Do Nothing
+        }
+        throw e
+    } finally {
+        try {
+            if (!closed)
+                scope?.close()
+            span.finish()
+        } catch (closeException: Exception) {
+            // Do Nothing
+        }
+    }
+}
+
+inline fun <R> Tracer.trace(
+    operation: String,
+    spanBuilder: Tracer.SpanBuilder.() -> Unit,
+    block: (Span) -> R
+): R {
+    val builder = buildSpan(operation)
+    builder.apply(spanBuilder)
+    val span = builder.start()
+    return trace(span, block)
+}
+
+inline fun <R> Tracer.trace(
+    operation: String,
+    component: String? = null,
+    kind: String? = null,
+    parent: SpanContext? = null,
+    block: (Span) -> R
+): R {
+    return trace(
+        operation = operation,
+        spanBuilder = {
+            parent?.let { asChildOf(parent) }
+            component?.let { withTag(Tags.COMPONENT, component) }
+            kind?.let { withTag(Tags.SPAN_KIND, kind) }
+        },
+        block = block
+    )
+}
+
+fun Span.record(ex: java.lang.Exception) {
+    log(
+        mapOf(
+            "event" to Tags.ERROR.key,
+            "error.object" to ex
+        )
+    )
+    setTag(Tags.ERROR, true)
+}
+
+fun Tracer.injectToMap(span: Span? = activeSpan()): Map<String, String>? =
```

Comment:
Even though the return type is just `Map`, you're still effectively returning a mutable map.

Reply:
yes, that is because of how the opentracing java library works.

Reply:
Not necessarily. The OpenTracing API just expects a `TextMapInject`. You picked the `TextMapAdapter` which itself requires a mutable map. But nobody forces that. You can have your own `TextMapInject` implementation that works with https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/build-map.html or just wrap the return value in `Collections.unmodifiableMap`

Reply:
why would we do that? Returning a MutableMap as Map is an accepted kotlin idiom.

Reply:
I was following up on the argument that passing in a mutable map is bad. I don't agree. But I don't see why returning a mutable map is better. Even if you return `Map`, it's still mutable.

Reply:
yes, it's still mutable if you cast it to a MutableMap - this serves as a big warning sign.
This is the whole idea of Map/MutableMap in kotlin.
Scala does it differently, where you could return an "ImmutableMap"...

Reply:
so in kotlin, `Map` is immutable always unless casted to `MutableMap`

Reply:
> unless casted

That's not what immutability means. If there is such an easy way to mutate X, then X is simply not immutable.
While it should be easy to create (i.e. copy, craft or build) a mutable map out of an immutable one, it shouldn't be possible to just simply treat an immutable map as mutable. Because that violate the very principle of immutability that users of an API rely upon.

Reply:
this is simply a kotlin idiom - read-only instead of true immutability

Reply:
> read-only

That makes a lot more sense as a term compared to _immutable_.